mlflow/java/client/src/main/java/org/mlflow/artifacts/CliBasedArtifactRepository.java

```diff
@@ -0,0 +1,242 @@
+package org.mlflow.artifacts;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import com.google.common.collect.Lists;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.util.JsonFormat;
+import org.apache.commons.io.IOUtils;
+import org.apache.log4j.Logger;
+
+import org.mlflow.api.proto.Service;
+import org.mlflow.tracking.MlflowClientException;
+import org.mlflow.tracking.creds.MlflowHostCredsProvider;
+
+/**
+ * Shells out to the 'mlflow' command line utility to upload, download, and list artifacts. This
+ * is used as a fallback to implement any artifact repositories which are not natively supported
+ * within Java.
+ *
+ * We require that 'mlflow' is available in the system path.
+ */
+public class CliBasedArtifactRepository implements ArtifactRepository {
+  private static final Logger logger = Logger.getLogger(CliBasedArtifactRepository.class);
+
+  // Global check if we ever successfully loaded 'mlflow'. This allows us to print a more
+  // helpful error message if the executable is not in the path.
+  private static final AtomicBoolean mlflowSuccessfullyLoaded = new AtomicBoolean(false);
+
+  // Name of the mlflow CLI utility which can be exec'd directly.
+  private final String mlflowExecutable = "mlflow";
+
+  // Base directory of the artifactory, used to let the user know why this repository was chosen.
+  private final String artifactBaseDir;
+
+  // Run ID this repository is targeting.
+  private final String runId;
```

Comment:
recommend calling this `runUuid` for disambiguation.

Reply:
I think we're trying to rename `runUuid` to `runId` everywhere per @mateiz's hope. The protobuf may be the last to go, though.